curvefs/src/metaserver/copyset/copyset_reloader.cpp

```diff
@@ -0,0 +1,254 @@
+      running_(false) {
+    if (options_.loadConcurrency > 0) {
+        taskPool_ = absl::make_unique<TaskThreadPool<>>();
+    }
+}
+
+bool CopysetReloader::ReloadCopysets() {
+    std::string dataDir = UriParser::GetPathFromUri(options_.dataUri);
+    if (dataDir.empty()) {
+        LOG(ERROR) << "Get path from uri may failed, data uri: '"
+                   << options_.dataUri << "'";
+        return false;
+    }
+
+    if (!options_.localFileSystem->DirExists(dataDir)) {
+        LOG(INFO) << "Data path '" << dataDir << "' not exist";
+        return true;
+    }
+
+    std::vector<std::string> copysets;
+    if (0 != options_.localFileSystem->List(dataDir, &copysets)) {
+        LOG(ERROR) << "Failed to get copyset list from data path " << dataDir;
+        return false;
+    }
+
+    LOG(INFO) << "Begin to load copysets under '" << dataDir << "'";
+    for (const auto& copyset : copysets) {
+        if (!ReloadOneCopyset(copyset)) {
+            LOG(ERROR) << "Reload copyset " << copyset
+                       << " failed, data path: " << dataDir;
+            return false;
+        }
+    }
+
+    WaitLoadFinish();
+
+    LOG(INFO) << "Load copysets under '" << dataDir << "' success";
+
+    return true;
+}
+
+bool CopysetReloader::ReloadOneCopyset(const std::string& copyset) {
+    LOG(INFO) << "Found copyset " << copyset;
+
+    GroupNid groupid = 0;
+    if (!curve::common::StringToUll(copyset, &groupid)) {
+        LOG(ERROR) << "Parse " << copyset << " to GroupNid failed";
+        return false;
+    }
+
+    PoolId poolId = GetPoolId(groupid);
+    CopysetId copysetId = GetCopysetId(groupid);
+
+    LOG(INFO) << "Parse " << copyset << " as "
+              << ToGroupIdString(poolId, copysetId);
+
+    if (taskPool_) {
+        taskPool_->Enqueue(&CopysetReloader::LoadCopyset, this, poolId,
+                           copysetId, true);
+    } else {
+        LoadCopyset(poolId, copysetId, false);
+    }
+
+    return true;
+}
+
+void CopysetReloader::LoadCopyset(PoolId poolId, CopysetId copysetId,
+                                  bool checkLoadFinish) {
+    LOG(INFO) << "Begin to load copyset: " << ToGroupIdString(poolId, copysetId)
+              << ", check load finish: " << std::boolalpha << checkLoadFinish;
+
+    auto begin = TimeUtility::GetTimeofDayMs();
+    braft::Configuration conf;
+    std::unique_ptr<CopysetNode> node =
+        CreateCopysetNode(poolId, copysetId, conf);
+
+    if (!node) {
+        LOG(ERROR) << "Failed to create copyset "
+                   << ToGroupIdString(poolId, copysetId);
+        return;
+    }
+
+    if (!nodeManager_->InsertCopysetNodeIfNotExist(poolId, copysetId,
+                                                   node.get())) {
+        LOG(ERROR) << "Failed to insert copyset "
+                   << ToGroupIdString(poolId, copysetId);
+        return;
+    }
+
+    // release ownership, node has been transffered to nodeManager
+    node.release();
+
+    if (checkLoadFinish) {
```

Comment:
if copysets are reloading concurrently, `checkLoadFinish` is true, because after a copysetnode/raftnode is created, it will recovers from raft snapshot/logs. so, if we don't wait copysets recover successfully, they can be recover concurrently, which will cause high cpu usage, high disk bandwidth